<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <title>Chapter 8 Limma analysis | Data Analysis in R</title>

    <meta name="author" content="Nico Hüttmann" />
  
   <meta name="description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::bs4_book,
set in the _output.yml file.</p>" />
   <meta name="generator" content="placeholder" />
  <meta property="og:title" content="Chapter 8 Limma analysis | Data Analysis in R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::bs4_book,
set in the _output.yml file.</p>" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Limma analysis | Data Analysis in R" />
  
  <meta name="twitter:description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::bs4_book,
set in the _output.yml file.</p>" />
  
  <!-- JS -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script>
  <script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script>
    <script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no" />
    <link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet" />
    <script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script>
    <script src="libs/bs3compat-0.7.0/transition.js"></script>
    <script src="libs/bs3compat-0.7.0/tabs.js"></script>
    <script src="libs/bs3compat-0.7.0/bs3compat.js"></script>
    <link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet" />
    <script src="libs/bs4_book-1.0.0/bs4_book.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script>

  <!-- CSS -->
  <style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
      <link rel="stylesheet" href="style.css" />
  
</head>

<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book">
    <a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Data Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
      </form>

      <nav aria-label="Table of contents">
        <h2>Table of contents</h2>
        <div id="book-toc"></div>

        <div class="book-extra">
          <p><a id="book-repo" href="#">View book source <i class="fab fa-github"></i></a></li></p>
        </div>
      </nav>
    </div>
  </header>

  <main class="col-sm-12 col-md-9 col-lg-7" id="content">
<div id="limma-analysis" class="section level1" number="8">
<h1><span class="header-section-number">Chapter 8</span> Limma analysis</h1>
<p>Here we are, after wrangling two days with our data, making it pretty and tidy without ever seeing a heatmap of it, ready to do a differential abundance analysis with <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC4402510/">limma</a>. <em>limma</em> is a tool originally developed for microarray analysis, which works just as well for proteomics data.<a href="#fn34" class="footnote-ref" id="fnref34"><sup>34</sup></a> We will use it to identify, which peptides have a different abundance between the two sample groups.</p>
<div id="running-limma" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Running limma</h2>
<p>You can find the <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma package</a> on Bioconductor. Check out the website and find 1. how to install it, 2. the limma User’s Guide and 3. the Reference Manual.<a href="#fn35" class="footnote-ref" id="fnref35"><sup>35</sup></a> All Bioconductor package pages are made up this way and allow us to find the relevant information quickly. Except in this case.</p>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button38" aria-expanded="false" aria-controls="button38">
Code + Links
</button>
<div id="button38" class="collapse">
<p><br />
</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="limma-analysis.html#cb74-1" tabindex="-1"></a><span class="co"># Install BiocManager if not yet installed</span></span>
<span id="cb74-2"><a href="limma-analysis.html#cb74-2" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb74-3"><a href="limma-analysis.html#cb74-3" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb74-4"><a href="limma-analysis.html#cb74-4" tabindex="-1"></a><span class="co"># Install limma</span></span>
<span id="cb74-5"><a href="limma-analysis.html#cb74-5" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;limma&quot;</span>)</span></code></pre></div>
<p><a href="https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf">limma User’s Guide</a> and <a href="https://bioconductor.org/packages/release/bioc/manuals/limma/man/limma.pdf">Reference Manual</a></p>
</div>
<p><br />
Admittedly, this is not very helpful yet. For other packages, you can find a much more concise documentation and commonly one or more workflow descriptions depending on what you want to do. <a href="https://kasperdanielhansen.github.io/genbioconductor/html/limma.html">This guide</a> seems a bit more practical and can be applied to our data.</p>
<div class="rmdnote">
<p>Let’s make this a real world example! Try and apply a <code>limma</code> analysis to your data. Use the explanations in <a href="https://kasperdanielhansen.github.io/genbioconductor/html/limma.html">this guide</a> and on page 36/37 of the <a href="https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf">limma User’s Guide</a>. Help for individual functions can be found in the <a href="https://bioconductor.org/packages/release/bioc/manuals/limma/man/limma.pdf">Reference Manual</a> and with the known <code>?limma::lmFit</code>.</p>
<p>The final goal is to use the <code>limma::topTable()</code> function to extract the results. Have fun and good luck!</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button39" aria-expanded="false" aria-controls="button39">
Solution
</button>
<div id="button39" class="collapse">
<p><br />
Are you sure you want to give up already?</p>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button40" aria-expanded="false" aria-controls="button40">
No
</button>
<div id="button40" class="collapse">
<p><br />
Great! Just continue trying out something and it will eventually work. In the worst case, you could use ChatG…NO, please don’t do this.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button41" aria-expanded="false" aria-controls="button41">
Yes
</button>
<div id="button41" class="collapse">
<p><br />
Okay, here we go. Since the analysis involves a couple of steps and I do not like to clutter the environment too much, I propose to keep intermediate steps of the same analysis all in one list. Don’t worry, the result will be the same as when you simply do it with multiple R objects.</p>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="limma-analysis.html#cb75-1" tabindex="-1"></a>limma_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span></code></pre></div>
<p>We will first add our data, which we already very nice prepared. The only thing missing is to take the <code>log2</code> of our data, because limma like other statistical tests like a t-test assume normally distributed data.</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="limma-analysis.html#cb76-1" tabindex="-1"></a><span class="do">## Add base data frame </span></span>
<span id="cb76-2"><a href="limma-analysis.html#cb76-2" tabindex="-1"></a>limma_list[[<span class="st">&quot;data&quot;</span>]] <span class="ot">&lt;-</span> data_norm <span class="sc">%&gt;%</span> </span>
<span id="cb76-3"><a href="limma-analysis.html#cb76-3" tabindex="-1"></a>  <span class="co"># This works simply with mutate(across(where(is.numeric), log2)) or with the </span></span>
<span id="cb76-4"><a href="limma-analysis.html#cb76-4" tabindex="-1"></a>  <span class="co"># fast version by pivoting in long format and back after the operation</span></span>
<span id="cb76-5"><a href="limma-analysis.html#cb76-5" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="sc">-</span><span class="dv">1</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb76-6"><a href="limma-analysis.html#cb76-6" tabindex="-1"></a>  <span class="co"># We need to make sure the sample sames stay in the correct order </span></span>
<span id="cb76-7"><a href="limma-analysis.html#cb76-7" tabindex="-1"></a>  <span class="fu">arrange</span>(name) <span class="sc">%&gt;%</span> </span>
<span id="cb76-8"><a href="limma-analysis.html#cb76-8" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">value =</span> <span class="fu">log2</span>(value)) <span class="sc">%&gt;%</span> </span>
<span id="cb76-9"><a href="limma-analysis.html#cb76-9" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>()</span></code></pre></div>
<p>It is good practise to save your input data close to your statistical tests.</p>
<p>The <code>limma</code> functions, similar to <code>limma::normalizeBetweenArrays()</code>, accept a <code>matrix</code> or an <code>eSet</code> (expression set), a special type of storing your data.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="limma-analysis.html#cb77-1" tabindex="-1"></a><span class="do">## Add eset </span></span>
<span id="cb77-2"><a href="limma-analysis.html#cb77-2" tabindex="-1"></a>limma_list[[<span class="st">&quot;eset&quot;</span>]] <span class="ot">&lt;-</span> limma_list[[<span class="st">&quot;data&quot;</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb77-3"><a href="limma-analysis.html#cb77-3" tabindex="-1"></a>  <span class="co"># We have seen this before, but here is the generalized example by renaming </span></span>
<span id="cb77-4"><a href="limma-analysis.html#cb77-4" tabindex="-1"></a>  <span class="co"># the first column to &quot;rowname&quot;, which is the default column to use as </span></span>
<span id="cb77-5"><a href="limma-analysis.html#cb77-5" tabindex="-1"></a>  <span class="co"># rownames for column_to_rownames</span></span>
<span id="cb77-6"><a href="limma-analysis.html#cb77-6" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">rowname =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb77-7"><a href="limma-analysis.html#cb77-7" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>() <span class="sc">%&gt;%</span></span>
<span id="cb77-8"><a href="limma-analysis.html#cb77-8" tabindex="-1"></a>  <span class="fu">as.matrix</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb77-9"><a href="limma-analysis.html#cb77-9" tabindex="-1"></a>  <span class="co"># Depending on your input data frame, you may need to transpose your data</span></span>
<span id="cb77-10"><a href="limma-analysis.html#cb77-10" tabindex="-1"></a>  <span class="co"># an eset has samples as columns and features/variables as rows</span></span>
<span id="cb77-11"><a href="limma-analysis.html#cb77-11" tabindex="-1"></a>  <span class="co"># t() %&gt;% </span></span>
<span id="cb77-12"><a href="limma-analysis.html#cb77-12" tabindex="-1"></a>  Biobase<span class="sc">::</span><span class="fu">ExpressionSet</span>()</span></code></pre></div>
<p>The ExpressionSet is an idea to standardize data formats in R, but is practically only used in some packages as input and most people mainly work with <code>tibble</code>s.</p>
<p>Next, we need to describe our experimental setup. We can find this information in our sample name <code>names(limma[["data"]])</code> and it describes either a ‘Control’ or the ‘Staurosporine’ group, we can simply make a vector that describes the sample groups in the order of our samples</p>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="limma-analysis.html#cb78-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;design&quot;</span>]] <span class="ot">&lt;-</span> </span>
<span id="cb78-2"><a href="limma-analysis.html#cb78-2" tabindex="-1"></a>  <span class="fu">model.matrix</span>(</span>
<span id="cb78-3"><a href="limma-analysis.html#cb78-3" tabindex="-1"></a>    <span class="sc">~</span><span class="dv">0</span><span class="sc">+</span><span class="fu">factor</span>(<span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">&quot;Control&quot;</span>, <span class="dv">4</span>), <span class="fu">rep</span>(<span class="st">&quot;Staurosporine&quot;</span>, <span class="dv">4</span>)), </span>
<span id="cb78-4"><a href="limma-analysis.html#cb78-4" tabindex="-1"></a>              <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">&quot;Control&quot;</span>, <span class="st">&quot;Staurosporine&quot;</span>)))</span></code></pre></div>
<p>The vector can be made in any way. <code>c(rep("Control", 4), rep("Staurosporine", 4))</code> is just one of many ways to accomplish this. <code>rep(c("Control", "Staurosporine"), each = 4)</code> would work as well or fully writing all names individually.</p>
<p>Because the names look weird, we can rename the generated model matrix. Have a look at it and see how they encode the group information in it.</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="limma-analysis.html#cb79-1" tabindex="-1"></a><span class="fu">colnames</span>(limma_list[[<span class="st">&quot;design&quot;</span>]]) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Control&quot;</span>, <span class="st">&quot;Staurosporine&quot;</span>)</span></code></pre></div>
<p>Now we can start with the first linear model fit <code>limma::lmFit()</code>. This needs the eset and the design matrix as input.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="limma-analysis.html#cb80-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;fit&quot;</span>]] <span class="ot">&lt;-</span> limma<span class="sc">::</span><span class="fu">lmFit</span>(limma_list[[<span class="st">&quot;eset&quot;</span>]], limma_list[[<span class="st">&quot;design&quot;</span>]])</span></code></pre></div>
<p>Following this, we need to indicate which comparisons we want to conduct. In the case of more than two groups or a more complex factorial design, we can indicate all interesting difference here.</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="limma-analysis.html#cb81-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;contrast.matrix&quot;</span>]] <span class="ot">&lt;-</span> limma<span class="sc">::</span><span class="fu">makeContrasts</span>(Staurosporine<span class="sc">-</span>Control, </span>
<span id="cb81-2"><a href="limma-analysis.html#cb81-2" tabindex="-1"></a>                                                        <span class="at">levels=</span>limma_list[[<span class="st">&quot;design&quot;</span>]])</span></code></pre></div>
<p>Now we compare both of our groups.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="limma-analysis.html#cb82-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;fit2&quot;</span>]] <span class="ot">&lt;-</span> limma<span class="sc">::</span><span class="fu">contrasts.fit</span>(<span class="at">fit =</span> limma_list[[<span class="st">&quot;fit&quot;</span>]], </span>
<span id="cb82-2"><a href="limma-analysis.html#cb82-2" tabindex="-1"></a>                                             <span class="at">contrasts =</span> limma_list[[<span class="st">&quot;contrast.matrix&quot;</span>]])</span></code></pre></div>
<p>And finally compute our statistics.</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="limma-analysis.html#cb83-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;fit2_eBayes&quot;</span>]] <span class="ot">&lt;-</span> limma<span class="sc">::</span><span class="fu">eBayes</span>(limma_list[[<span class="st">&quot;fit2&quot;</span>]])</span></code></pre></div>
<p>The final results can be extracted with <code>limma::topTable()</code>.</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="limma-analysis.html#cb84-1" tabindex="-1"></a>limma<span class="sc">::</span><span class="fu">topTable</span>(limma_list[[<span class="st">&quot;fit2_eBayes&quot;</span>]]) <span class="co"># %&gt;% View()</span></span></code></pre></div>
<p>Great, we go our first results. Let’s explore how we can further work with them.</p>
</div>
</div>
</div>
<div id="filtering-the-limma-output" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Filtering the limma output</h2>
<p>The <code>limma::topTable()</code> should have given you a preview of your results. But I recommend using a different functions for this, the <code>biobroom::tidy.MArrayLM()</code> function which outputs a nice <code>tibble</code>. <a href="https://bioconductor.org/packages/release/bioc/html/biobroom.html">biobroom</a> is again a Bioconductor function and it helps you with “[…] methods for converting standard objects constructed by bioinformatics packages, especially those in Bioconductor, and converting them to tidy data”.</p>
<div class="rmdnote">
<p>Try it out and have a look at your results!<a href="#fn36" class="footnote-ref" id="fnref36"><sup>36</sup></a></p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button42" aria-expanded="false" aria-controls="button42">
Code
</button>
<div id="button42" class="collapse">
<p><br />
</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="limma-analysis.html#cb85-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="ot">&lt;-</span> </span>
<span id="cb85-2"><a href="limma-analysis.html#cb85-2" tabindex="-1"></a>  biobroom<span class="sc">::</span><span class="fu">tidy.MArrayLM</span>(limma_list[[<span class="st">&quot;fit2_eBayes&quot;</span>]])</span></code></pre></div>
</div>
<p><br />
We can <code>View()</code> our results and bring the peptides with the lowest p-value to the top by clicking the <code>p.value</code> column. Still, there are still some generic terms in this data frame like <code>gene</code>. We can modify it to help its legibility.</p>
<div class="rmdnote">
<p>Repeat the last command but further modify the data frame by renaming the gene column to something like id and arrange the rows by p.value.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button43" aria-expanded="false" aria-controls="button43">
Code
</button>
<div id="button43" class="collapse">
<p><br />
</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="limma-analysis.html#cb86-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="ot">&lt;-</span> biobroom<span class="sc">::</span><span class="fu">tidy.MArrayLM</span>(limma_list[[<span class="st">&quot;fit2_eBayes&quot;</span>]]) <span class="sc">%&gt;%</span> </span>
<span id="cb86-2"><a href="limma-analysis.html#cb86-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">id =</span> gene) <span class="sc">%&gt;%</span> </span>
<span id="cb86-3"><a href="limma-analysis.html#cb86-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(p.value)</span></code></pre></div>
</div>
<p><br />
Nice, now we can look at the data more closely. A common practise when applying statistical test to many different measurements is a multiple testing correction, as even with a p-value threshold of 0.05, we assume 5% of our hits to be false positives. The most common method in proteomics is the Benjamini-Hochberg correction, which is implemented in the <code>p.adjust()</code> function.</p>
<div class="rmdnote">
<p>Add a <code>p.adjust</code> column to your data by applying the <code>p.adjust()</code> function to your p-values.<a href="#fn37" class="footnote-ref" id="fnref37"><sup>37</sup></a> Typing <code>p.adjust.methods</code> will tell you about available methods. You can again append the code from above.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button44" aria-expanded="false" aria-controls="button44">
Code
</button>
<div id="button44" class="collapse">
<p><br />
</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="limma-analysis.html#cb87-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="ot">&lt;-</span> biobroom<span class="sc">::</span><span class="fu">tidy.MArrayLM</span>(limma_list[[<span class="st">&quot;fit2_eBayes&quot;</span>]]) <span class="sc">%&gt;%</span> </span>
<span id="cb87-2"><a href="limma-analysis.html#cb87-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">id =</span> gene) <span class="sc">%&gt;%</span> </span>
<span id="cb87-3"><a href="limma-analysis.html#cb87-3" tabindex="-1"></a>  <span class="co"># Rearranging our data by the p-values does not affect the adjustment </span></span>
<span id="cb87-4"><a href="limma-analysis.html#cb87-4" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(p.value) <span class="sc">%&gt;%</span> </span>
<span id="cb87-5"><a href="limma-analysis.html#cb87-5" tabindex="-1"></a>  <span class="co"># Adjust p-values</span></span>
<span id="cb87-6"><a href="limma-analysis.html#cb87-6" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">p.adjust =</span> <span class="fu">p.adjust</span>(p.value, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>), </span>
<span id="cb87-7"><a href="limma-analysis.html#cb87-7" tabindex="-1"></a>                <span class="at">.after =</span> <span class="st">&quot;p.value&quot;</span>)</span></code></pre></div>
</div>
<p><br />
Nice! Finally, we want to use our adjusted p-values and the effect size/log2 fold-change (FC) termed <code>estimate</code> in the table, to threshold our data and define more and less abundant/up- and down-regulated peptides. We can do this with the <code>dplyr::case_when()</code> function. Have a look at the code!</p>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button45" aria-expanded="false" aria-controls="button45">
Code
</button>
<div id="button45" class="collapse">
<p><br />
</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="limma-analysis.html#cb88-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="ot">&lt;-</span> biobroom<span class="sc">::</span><span class="fu">tidy.MArrayLM</span>(limma_list[[<span class="st">&quot;fit2_eBayes&quot;</span>]]) <span class="sc">%&gt;%</span> </span>
<span id="cb88-2"><a href="limma-analysis.html#cb88-2" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">id =</span> gene) <span class="sc">%&gt;%</span> </span>
<span id="cb88-3"><a href="limma-analysis.html#cb88-3" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">arrange</span>(p.value) <span class="sc">%&gt;%</span> </span>
<span id="cb88-4"><a href="limma-analysis.html#cb88-4" tabindex="-1"></a>  <span class="co"># Adjust p-values</span></span>
<span id="cb88-5"><a href="limma-analysis.html#cb88-5" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">p.adjust =</span> <span class="fu">p.adjust</span>(p.value, <span class="at">method =</span> <span class="st">&quot;BH&quot;</span>), </span>
<span id="cb88-6"><a href="limma-analysis.html#cb88-6" tabindex="-1"></a>                <span class="at">.after =</span> <span class="st">&quot;p.value&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb88-7"><a href="limma-analysis.html#cb88-7" tabindex="-1"></a>  <span class="co"># Apply thresholds</span></span>
<span id="cb88-8"><a href="limma-analysis.html#cb88-8" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">regulation =</span> dplyr<span class="sc">::</span><span class="fu">case_when</span>(</span>
<span id="cb88-9"><a href="limma-analysis.html#cb88-9" tabindex="-1"></a>    p.adjust <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">&amp;</span> </span>
<span id="cb88-10"><a href="limma-analysis.html#cb88-10" tabindex="-1"></a>      estimate <span class="sc">&gt;=</span> <span class="fu">log10</span>(<span class="fl">1.2</span>) <span class="sc">~</span> <span class="st">&quot;up&quot;</span>, </span>
<span id="cb88-11"><a href="limma-analysis.html#cb88-11" tabindex="-1"></a>    p.adjust <span class="sc">&lt;</span> <span class="fl">0.01</span> <span class="sc">&amp;</span> </span>
<span id="cb88-12"><a href="limma-analysis.html#cb88-12" tabindex="-1"></a>      estimate <span class="sc">&lt;=</span> <span class="sc">-</span> <span class="fu">log10</span>(<span class="fl">1.2</span>) <span class="sc">~</span> <span class="st">&quot;down&quot;</span>, </span>
<span id="cb88-13"><a href="limma-analysis.html#cb88-13" tabindex="-1"></a>    <span class="at">.default =</span> <span class="st">&quot;none&quot;</span></span>
<span id="cb88-14"><a href="limma-analysis.html#cb88-14" tabindex="-1"></a>  ))</span></code></pre></div>
<p>Each logical statement like <code>p.adjust &lt; 0.01 &amp; estimate &gt;= log10(1.2)</code><a href="#fn38" class="footnote-ref" id="fnref38"><sup>38</sup></a> return if true the values defined after <code>~</code>. If none apply, the function will return the <code>.default</code> for this element.<a href="#fn39" class="footnote-ref" id="fnref39"><sup>39</sup></a></p>
</div>
<p><br />
The table in <code>limma_list[["results"]]</code> contains now all information we need to further plot and report our results.</p>
</div>
<div id="visualizing-the-limma-results" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> Visualizing the limma results</h2>
<p>The one and only way to represent the data from a differential abundance/expression analysis with two groups is the Volcano plot. Let’s try it!</p>
<div class="rmdnote">
<p>Represent the <em>limma</em> results <code>estimate</code> and the <code>-log10(p.value)</code> in a volcano plot and color each point by the type of <code>regulation</code>. Use <code>ggplot2</code> for this.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button46" aria-expanded="false" aria-controls="button46">
Expanation
</button>
<div id="button46" class="collapse">
<p><br />
The <code>ggplot2::ggplot()</code> function builds the foundation of your plot. You can try this function without any further input.</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="limma-analysis.html#cb89-1" tabindex="-1"></a><span class="fu">ggplot</span>()</span></code></pre></div>
<p>You should see a blank canvas. We can start to add our data with the <code>data</code> argument.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="limma-analysis.html#cb90-1" tabindex="-1"></a><span class="fu">ggplot</span>(limma_list[[<span class="st">&quot;results&quot;</span>]]) </span></code></pre></div>
<p>Wow, nothing changed. Next, we need to explain <code>ggplot</code> which column of our <code>data</code> should be projected on which axis or plot element. This is achieved with the <code>mapping</code> argument which takes the above described links between data and plot within the <code>ggplot2::aes()</code><a href="#fn40" class="footnote-ref" id="fnref40"><sup>40</sup></a> function. We can define columns like <code>x = estimate</code> and even modify them in place like <code>y = -log10(p.value)</code>.</p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="limma-analysis.html#cb91-1" tabindex="-1"></a><span class="fu">ggplot</span>(limma_list[[<span class="st">&quot;results&quot;</span>]], <span class="fu">aes</span>(<span class="at">x =</span> estimate, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value)))</span></code></pre></div>
<p>Now that the call gets longer, we can make use of the pipe to add our data.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="limma-analysis.html#cb92-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb92-2"><a href="limma-analysis.html#cb92-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> estimate, <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value)))</span></code></pre></div>
<p>Oh, we have some axes now. But <code>ggplot</code> still only knows which data should be represented where, hence the axes, but not how. We want the most basic type, points.</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="limma-analysis.html#cb93-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb93-2"><a href="limma-analysis.html#cb93-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> estimate,</span>
<span id="cb93-3"><a href="limma-analysis.html#cb93-3" tabindex="-1"></a>             <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value))) <span class="sc">+</span></span>
<span id="cb93-4"><a href="limma-analysis.html#cb93-4" tabindex="-1"></a>  <span class="co"># geom_point is only one of many geom_... layers </span></span>
<span id="cb93-5"><a href="limma-analysis.html#cb93-5" tabindex="-1"></a>  <span class="fu">geom_point</span>()</span></code></pre></div>
<p>Nice, each one of our precious peptides represented as one little ugly point. Let’s unravel this. <code>ggplot2::geom_point()</code> is called a layer and adds the representation for the defined <code>mapping</code> of our data. It is added to the initial plot created by <code>ggplot2::ggplot()</code> via a <code>+</code>. No <code>%&gt;%</code>. Just a <code>+</code>. We can further modify this layer by setting a transparency with <code>alpha</code> and changing the <code>shape</code>, so that the changes are better visible.<a href="#fn41" class="footnote-ref" id="fnref41"><sup>41</sup></a></p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="limma-analysis.html#cb94-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb94-2"><a href="limma-analysis.html#cb94-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> estimate,</span>
<span id="cb94-3"><a href="limma-analysis.html#cb94-3" tabindex="-1"></a>             <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value))) <span class="sc">+</span></span>
<span id="cb94-4"><a href="limma-analysis.html#cb94-4" tabindex="-1"></a>  <span class="co"># you can check all options in the function documentation </span></span>
<span id="cb94-5"><a href="limma-analysis.html#cb94-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p>Further, we can change the <code>theme</code> to something like <code>ggplot2::theme_classic()</code>.</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="limma-analysis.html#cb95-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb95-2"><a href="limma-analysis.html#cb95-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> estimate,</span>
<span id="cb95-3"><a href="limma-analysis.html#cb95-3" tabindex="-1"></a>             <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value))) <span class="sc">+</span></span>
<span id="cb95-4"><a href="limma-analysis.html#cb95-4" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb95-5"><a href="limma-analysis.html#cb95-5" tabindex="-1"></a>  <span class="co"># there are many different theme_... types </span></span>
<span id="cb95-6"><a href="limma-analysis.html#cb95-6" tabindex="-1"></a>  <span class="fu">theme_classic</span>()</span></code></pre></div>
<p>Looking good! Now let’s color our points. If we would like to color our points all in one color, we could simple define this in the <code>ggplot2::geom_point(color = "grey)</code> layer. But we want the color to depend on the type of regulation as defined by the thresholds before. For this, we need to add a new variable to the <code>aes()</code> mapping and define a manual color scale.</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="limma-analysis.html#cb96-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb96-2"><a href="limma-analysis.html#cb96-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> estimate,</span>
<span id="cb96-3"><a href="limma-analysis.html#cb96-3" tabindex="-1"></a>             <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value),</span>
<span id="cb96-4"><a href="limma-analysis.html#cb96-4" tabindex="-1"></a>             <span class="co"># Define which column to use</span></span>
<span id="cb96-5"><a href="limma-analysis.html#cb96-5" tabindex="-1"></a>             <span class="at">color =</span> regulation)) <span class="sc">+</span></span>
<span id="cb96-6"><a href="limma-analysis.html#cb96-6" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb96-7"><a href="limma-analysis.html#cb96-7" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb96-8"><a href="limma-analysis.html#cb96-8" tabindex="-1"></a>  <span class="co"># Define the values for each data entry</span></span>
<span id="cb96-9"><a href="limma-analysis.html#cb96-9" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="at">none =</span> <span class="st">&quot;grey&quot;</span>,</span>
<span id="cb96-10"><a href="limma-analysis.html#cb96-10" tabindex="-1"></a>                                <span class="at">up =</span> <span class="st">&quot;red&quot;</span>,</span>
<span id="cb96-11"><a href="limma-analysis.html#cb96-11" tabindex="-1"></a>                                <span class="at">down =</span> <span class="st">&quot;blue&quot;</span>))</span></code></pre></div>
<p>Finally, we can remove the space between our plotted points and the outer border like this. This is a taste thing.</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="limma-analysis.html#cb97-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb97-2"><a href="limma-analysis.html#cb97-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> estimate,</span>
<span id="cb97-3"><a href="limma-analysis.html#cb97-3" tabindex="-1"></a>             <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value),</span>
<span id="cb97-4"><a href="limma-analysis.html#cb97-4" tabindex="-1"></a>             <span class="at">color =</span> regulation)) <span class="sc">+</span></span>
<span id="cb97-5"><a href="limma-analysis.html#cb97-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb97-6"><a href="limma-analysis.html#cb97-6" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb97-7"><a href="limma-analysis.html#cb97-7" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="at">none =</span> <span class="st">&quot;grey&quot;</span>,</span>
<span id="cb97-8"><a href="limma-analysis.html#cb97-8" tabindex="-1"></a>                                <span class="at">up =</span> <span class="st">&quot;red&quot;</span>,</span>
<span id="cb97-9"><a href="limma-analysis.html#cb97-9" tabindex="-1"></a>                                <span class="at">down =</span> <span class="st">&quot;blue&quot;</span>)) <span class="sc">+</span></span>
<span id="cb97-10"><a href="limma-analysis.html#cb97-10" tabindex="-1"></a>  <span class="co"># You can also define the xlim = c(-3, 3) and ylim = c(0, 10) here </span></span>
<span id="cb97-11"><a href="limma-analysis.html#cb97-11" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">expand =</span> F)</span></code></pre></div>
<p>Finally, we can save our plot first and print it then separately.</p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="limma-analysis.html#cb98-1" tabindex="-1"></a>limma_list[[<span class="st">&quot;p&quot;</span>]] <span class="ot">&lt;-</span> limma_list[[<span class="st">&quot;results&quot;</span>]] <span class="sc">%&gt;%</span></span>
<span id="cb98-2"><a href="limma-analysis.html#cb98-2" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> estimate,</span>
<span id="cb98-3"><a href="limma-analysis.html#cb98-3" tabindex="-1"></a>             <span class="at">y =</span> <span class="sc">-</span><span class="fu">log10</span>(p.value),</span>
<span id="cb98-4"><a href="limma-analysis.html#cb98-4" tabindex="-1"></a>             <span class="at">color =</span> regulation)) <span class="sc">+</span></span>
<span id="cb98-5"><a href="limma-analysis.html#cb98-5" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">shape =</span> <span class="dv">16</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb98-6"><a href="limma-analysis.html#cb98-6" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb98-7"><a href="limma-analysis.html#cb98-7" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="at">none =</span> <span class="st">&quot;grey&quot;</span>,</span>
<span id="cb98-8"><a href="limma-analysis.html#cb98-8" tabindex="-1"></a>                                <span class="at">up =</span> <span class="st">&quot;red&quot;</span>,</span>
<span id="cb98-9"><a href="limma-analysis.html#cb98-9" tabindex="-1"></a>                                <span class="at">down =</span> <span class="st">&quot;blue&quot;</span>)) <span class="sc">+</span></span>
<span id="cb98-10"><a href="limma-analysis.html#cb98-10" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">expand =</span> F)</span>
<span id="cb98-11"><a href="limma-analysis.html#cb98-11" tabindex="-1"></a><span class="co"># Sometimes you may need force this by print(limma_list[[&quot;p&quot;]]) </span></span>
<span id="cb98-12"><a href="limma-analysis.html#cb98-12" tabindex="-1"></a>limma_list[[<span class="st">&quot;p&quot;</span>]]</span></code></pre></div>
</div>
<p><br />
Great, this basically covered most parameters you can change for plotting in R with <code>ggplot2</code>. It may take some time to get comfortable with all functions and get to know all possibilities, but then there are no limits!</p>
</div>
<div id="summary-1" class="section level2" number="8.4">
<h2><span class="header-section-number">8.4</span> Summary</h2>
<p>In this chapter, we applied a package <code>limma</code> to analyse our data and represented the results in one plot. <code>limma</code> is one of the more complicated packages to use IMO, but still very useful and a good exercise.</p>
</div>
</div>
<div class="footnotes">
<hr />
<ol start="34">
<li id="fn34"><p>Yes, proteomics started by borrowing tools from the RNA people..<a href="limma-analysis.html#fnref34" class="footnote-back">↩︎</a></p></li>
<li id="fn35"><p>You can leave out ‘A brief introduction to limma’<a href="limma-analysis.html#fnref35" class="footnote-back">↩︎</a></p></li>
<li id="fn36"><p>And install biobroom if you do not have it yet. <a href="limma-analysis.html#fnref36" class="footnote-back">↩︎</a></p></li>
<li id="fn37"><p>Bonus if the new column lands right after p.value.<a href="limma-analysis.html#fnref37" class="footnote-back">↩︎</a></p></li>
<li id="fn38"><p>log10(120%) refers to an increase of 20% in peptide abundance.<a href="limma-analysis.html#fnref38" class="footnote-back">↩︎</a></p></li>
<li id="fn39"><p>Which is only by chance called “none” here as well.<a href="limma-analysis.html#fnref39" class="footnote-back">↩︎</a></p></li>
<li id="fn40"><p>aesthetics<a href="limma-analysis.html#fnref40" class="footnote-back">↩︎</a></p></li>
<li id="fn41"><p>You can find a list of all shapes here: <a href="https://ggplot2.tidyverse.org/articles/ggplot2-specs.html#sec:shape-spec" class="uri">https://ggplot2.tidyverse.org/articles/ggplot2-specs.html#sec:shape-spec</a><a href="limma-analysis.html#fnref41" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </main>

  <div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page">
      <h2>On this page</h2>
      <div id="book-on-this-page"></div>

      <div class="book-extra">
        <ul class="list-unstyled">
          <li><a id="book-source" href="#">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="#">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
      </div>
    </nav>
  </div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5">
  <div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Data Analysis in R</strong>" was written by Nico Hüttmann. It was last built on 2025-02-06.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>


</body>

</html>
