<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 8 Limma analysis | Data Analysis in R</title>
<meta name="author" content="Nico Hüttmann">
<meta name="description" content="Here we are, after wrangling two days with our data, making it pretty and tidy without ever seeing a heatmap of it, ready to do a differential abundance analysis with limma. limma is a tool...">
<meta name="generator" content="bookdown 0.38 with bs4_book()">
<meta property="og:title" content="Chapter 8 Limma analysis | Data Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="Here we are, after wrangling two days with our data, making it pretty and tidy without ever seeing a heatmap of it, ready to do a differential abundance analysis with limma. limma is a tool...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 8 Limma analysis | Data Analysis in R">
<meta name="twitter:description" content="Here we are, after wrangling two days with our data, making it pretty and tidy without ever seeing a heatmap of it, ready to do a differential abundance analysis with limma. limma is a tool...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.7.0/transition.js"></script><script src="libs/bs3compat-0.7.0/tabs.js"></script><script src="libs/bs3compat-0.7.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Data Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">Preface</a></li>
<li class="book-part">R Basics</li>
<li><a class="" href="r-and-rstudio.html"><span class="header-section-number">1</span> R and RStudio</a></li>
<li><a class="" href="how-does-r-work.html"><span class="header-section-number">2</span> How does R work?</a></li>
<li><a class="" href="r-packages.html"><span class="header-section-number">3</span> R packages</a></li>
<li><a class="" href="the-tidyverse.html"><span class="header-section-number">4</span> The tidyverse</a></li>
<li><a class="" href="coding-style.html"><span class="header-section-number">5</span> Coding style</a></li>
<li class="book-part">Group Work Project</li>
<li><a class="" href="our-project.html"><span class="header-section-number">6</span> Our project</a></li>
<li><a class="" href="working-with-data.html"><span class="header-section-number">7</span> Working with data</a></li>
<li><a class="active" href="limma-analysis.html"><span class="header-section-number">8</span> Limma analysis</a></li>
<li><a class="" href="reporting-results.html"><span class="header-section-number">9</span> Reporting results</a></li>
<li><a class="" href="more-data-exploration.html"><span class="header-section-number">10</span> More data exploration</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="version-control-with-github.html"><span class="header-section-number">A</span> Version control with GitHub</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/nicohuttmann/embl-bioinfo-pRoteomics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="limma-analysis" class="section level1" number="8">
<h1>
<span class="header-section-number">8</span> Limma analysis<a class="anchor" aria-label="anchor" href="#limma-analysis"><i class="fas fa-link"></i></a>
</h1>
<p>Here we are, after wrangling two days with our data, making it pretty and tidy without ever seeing a heatmap of it, ready to do a differential abundance analysis with <a href="https://pmc.ncbi.nlm.nih.gov/articles/PMC4402510/">limma</a>. <em>limma</em> is a tool originally developed for microarray analysis, which works just as well for proteomics data.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Yes, proteomics started by borrowing tools from the RNA people..&lt;/p&gt;"><sup>34</sup></a> We will use it to identify, which peptides have a different abundance between the two sample groups.</p>
<div id="running-limma" class="section level2" number="8.1">
<h2>
<span class="header-section-number">8.1</span> Running limma<a class="anchor" aria-label="anchor" href="#running-limma"><i class="fas fa-link"></i></a>
</h2>
<p>You can find the <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma package</a> on Bioconductor. Check out the website and find 1. how to install it, 2. the limma User’s Guide and 3. the Reference Manual.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;You can leave out ‘A brief introduction to limma’&lt;/p&gt;"><sup>35</sup></a> All Bioconductor package pages are made up this way and allow us to find the relevant information quickly. Except in this case.</p>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button38" aria-expanded="false" aria-controls="button38">
Code + Links
</button>
<div id="button38" class="collapse">
<p><br></p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install BiocManager if not yet installed</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="co"># Install limma</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html">install</a></span><span class="op">(</span><span class="st">"limma"</span><span class="op">)</span></span></code></pre></div>
<p><a href="https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf">limma User’s Guide</a> and <a href="https://bioconductor.org/packages/release/bioc/manuals/limma/man/limma.pdf">Reference Manual</a></p>
</div>
<p><br>
Admittedly, this is not very helpful yet. For other packages, you can find a much more concise documentation and commonly one or more workflow descriptions depending on what you want to do. <a href="https://kasperdanielhansen.github.io/genbioconductor/html/limma.html">This guide</a> seems a bit more practical and can be applied to our data.</p>
<div class="rmdnote">
<p>Let’s make this a real world example! Try and apply a <code>limma</code> analysis to your data. Use the explanations in <a href="https://kasperdanielhansen.github.io/genbioconductor/html/limma.html">this guide</a> and on page 36/37 of the <a href="https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf">limma User’s Guide</a>. Help for individual functions can be found in the <a href="https://bioconductor.org/packages/release/bioc/manuals/limma/man/limma.pdf">Reference Manual</a> and with the known <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html">?limma::lmFit</a></code>.</p>
<p>The final goal is to use the <code><a href="https://rdrr.io/pkg/limma/man/toptable.html">limma::topTable()</a></code> function to extract the results. Have fun and good luck!</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button39" aria-expanded="false" aria-controls="button39">
Solution
</button>
<div id="button39" class="collapse">
<p><br>
Are you sure you want to give up already?</p>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button40" aria-expanded="false" aria-controls="button40">
No
</button>
<div id="button40" class="collapse">
<p><br>
Great! Just continue trying out something and it will eventually work. In the worst case, you could use ChatG…NO, please don’t do this.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button41" aria-expanded="false" aria-controls="button41">
Yes
</button>
<div id="button41" class="collapse">
<p><br>
Okay, here we go. Since the analysis involves a couple of steps and I do not like to clutter the environment too much, I propose to keep intermediate steps of the same analysis all in one list. Don’t worry, the result will be the same as when you simply do it with multiple R objects.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We will first add our data, which we already very nice prepared. The only thing missing is to take the <code>log2</code> of our data, because limma like other statistical tests like a t-test assume normally distributed data.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Add base data frame </span></span>
<span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">data_norm</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># This works simply with mutate(across(where(is.numeric), log2)) or with the </span></span>
<span>  <span class="co"># fast version by pivoting in long format and back after the operation</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># We need to make sure the sample sames stay in the correct order </span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>value <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">pivot_wider</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>It is good practise to save your input data close to your statistical tests.</p>
<p>The <code>limma</code> functions, similar to <code><a href="https://rdrr.io/pkg/limma/man/normalizebetweenarrays.html">limma::normalizeBetweenArrays()</a></code>, accept a <code>matrix</code> or an <code>eSet</code> (expression set), a special type of storing your data.</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Add eset </span></span>
<span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"eset"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">limma_list</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># We have seen this before, but here is the generalized example by renaming </span></span>
<span>  <span class="co"># the first column to "rowname", which is the default column to use as </span></span>
<span>  <span class="co"># rownames for column_to_rownames</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>rowname <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">column_to_rownames</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># Depending on your input data frame, you may need to transpose your data</span></span>
<span>  <span class="co"># an eset has samples as columns and features/variables as rows</span></span>
<span>  <span class="co"># t() %&gt;% </span></span>
<span>  <span class="fu">Biobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html">ExpressionSet</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The ExpressionSet is an idea to standardize data formats in R, but is practically only used in some packages as input and most people mainly work with <code>tibble</code>s.</p>
<p>Next, we need to describe our experimental setup. We can find this information in our sample name <code>names(limma[["data"]])</code> and it describes either a ‘Control’ or the ‘Staurosporine’ group, we can simply make a vector that describes the sample groups in the order of our samples</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"design"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span></span>
<span>    <span class="op">~</span><span class="fl">0</span><span class="op">+</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="fl">4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">"Staurosporine"</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span>, </span>
<span>              levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Staurosporine"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The vector can be made in any way. <code>c(rep("Control", 4), rep("Staurosporine", 4))</code> is just one of many ways to accomplish this. <code>rep(c("Control", "Staurosporine"), each = 4)</code> would work as well or fully writing all names individually.</p>
<p>Because the names look weird, we can rename the generated model matrix. Have a look at it and see how they encode the group information in it.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"design"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Staurosporine"</span><span class="op">)</span></span></code></pre></div>
<p>Now we can start with the first linear model fit <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html">limma::lmFit()</a></code>. This needs the eset and the design matrix as input.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html">lmFit</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"eset"</span><span class="op">]</span><span class="op">]</span>, <span class="va">limma_list</span><span class="op">[[</span><span class="st">"design"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Following this, we need to indicate which comparisons we want to conduct. In the case of more than two groups or a more complex factorial design, we can indicate all interesting difference here.</p>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"contrast.matrix"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html">makeContrasts</a></span><span class="op">(</span><span class="va">Staurosporine</span><span class="op">-</span><span class="va">Control</span>, </span>
<span>                                                        levels<span class="op">=</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"design"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Now we compare both of our groups.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit2"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/contrasts.fit.html">contrasts.fit</a></span><span class="op">(</span>fit <span class="op">=</span> <span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit"</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                             contrasts <span class="op">=</span> <span class="va">limma_list</span><span class="op">[[</span><span class="st">"contrast.matrix"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>And finally compute our statistics.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit2_eBayes"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit2"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>The final results can be extracted with <code><a href="https://rdrr.io/pkg/limma/man/toptable.html">limma::topTable()</a></code>.</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/toptable.html">topTable</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit2_eBayes"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># %&gt;% View()</span></span></code></pre></div>
<p>Great, we go our first results. Let’s explore how we can further work with them.</p>
</div>
</div>
</div>
<div id="filtering-the-limma-output" class="section level2" number="8.2">
<h2>
<span class="header-section-number">8.2</span> Filtering the limma output<a class="anchor" aria-label="anchor" href="#filtering-the-limma-output"><i class="fas fa-link"></i></a>
</h2>
<p>The <code><a href="https://rdrr.io/pkg/limma/man/toptable.html">limma::topTable()</a></code> should have given you a preview of your results. But I recommend using a different functions for this, the <code><a href="https://rdrr.io/pkg/biobroom/man/limma_tidiers.html">biobroom::tidy.MArrayLM()</a></code> function which outputs a nice <code>tibble</code>. <a href="https://bioconductor.org/packages/release/bioc/html/biobroom.html">biobroom</a> is again a Bioconductor function and it helps you with “[…] methods for converting standard objects constructed by bioinformatics packages, especially those in Bioconductor, and converting them to tidy data”.</p>
<div class="rmdnote">
<p>Try it out and have a look at your results!<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;And install biobroom if you do not have it yet. &lt;/p&gt;"><sup>36</sup></a></p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button42" aria-expanded="false" aria-controls="button42">
Code
</button>
<div id="button42" class="collapse">
<p><br></p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu">biobroom</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biobroom/man/limma_tidiers.html">tidy.MArrayLM</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit2_eBayes"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<p><br>
We can <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code> our results and bring the peptides with the lowest p-value to the top by clicking the <code>p.value</code> column. Still, there are still some generic terms in this data frame like <code>gene</code>. We can modify it to help its legibility.</p>
<div class="rmdnote">
<p>Repeat the last command but further modify the data frame by renaming the gene column to something like id and arrange the rows by p.value.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button43" aria-expanded="false" aria-controls="button43">
Code
</button>
<div id="button43" class="collapse">
<p><br></p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biobroom</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biobroom/man/limma_tidiers.html">tidy.MArrayLM</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit2_eBayes"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>id <span class="op">=</span> <span class="va">gene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span></span></code></pre></div>
</div>
<p><br>
Nice, now we can look at the data more closely. A common practise when applying statistical test to many different measurements is a multiple testing correction, as even with a p-value threshold of 0.05, we assume 5% of our hits to be false positives. The most common method in proteomics is the Benjamini-Hochberg correction, which is implemented in the <code><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust()</a></code> function.</p>
<div class="rmdnote">
<p>Add a <code>p.adjust</code> column to your data by applying the <code><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust()</a></code> function to your p-values.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Bonus if the new column lands right after p.value.&lt;/p&gt;"><sup>37</sup></a> Typing <code>p.adjust.methods</code> will tell you about available methods. You can again append the code from above.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button44" aria-expanded="false" aria-controls="button44">
Code
</button>
<div id="button44" class="collapse">
<p><br></p>
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biobroom</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biobroom/man/limma_tidiers.html">tidy.MArrayLM</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit2_eBayes"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>id <span class="op">=</span> <span class="va">gene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># Rearranging our data by the p-values does not affect the adjustment </span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># Adjust p-values</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>p.adjust <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">p.value</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span>, </span>
<span>                .after <span class="op">=</span> <span class="st">"p.value"</span><span class="op">)</span></span></code></pre></div>
</div>
<p><br>
Nice! Finally, we want to use our adjusted p-values and the effect size/log2 fold-change (FC) termed <code>estimate</code> in the table, to threshold our data and define more and less abundant/up- and down-regulated peptides. We can do this with the <code><a href="https://dplyr.tidyverse.org/reference/case_when.html">dplyr::case_when()</a></code> function. Have a look at the code!</p>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button45" aria-expanded="false" aria-controls="button45">
Code
</button>
<div id="button45" class="collapse">
<p><br></p>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">biobroom</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/biobroom/man/limma_tidiers.html">tidy.MArrayLM</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"fit2_eBayes"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>id <span class="op">=</span> <span class="va">gene</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># Adjust p-values</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>p.adjust <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html">p.adjust</a></span><span class="op">(</span><span class="va">p.value</span>, method <span class="op">=</span> <span class="st">"BH"</span><span class="op">)</span>, </span>
<span>                .after <span class="op">=</span> <span class="st">"p.value"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># Apply thresholds</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>regulation <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>    <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> </span>
<span>      <span class="va">estimate</span> <span class="op">&gt;=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">1.2</span><span class="op">)</span> <span class="op">~</span> <span class="st">"up"</span>, </span>
<span>    <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> </span>
<span>      <span class="va">estimate</span> <span class="op">&lt;=</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="fl">1.2</span><span class="op">)</span> <span class="op">~</span> <span class="st">"down"</span>, </span>
<span>    .default <span class="op">=</span> <span class="st">"none"</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Each logical statement like <code>p.adjust &lt; 0.01 &amp; estimate &gt;= log10(1.2)</code><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;log10(120%) refers to an increase of 20% in peptide abundance.&lt;/p&gt;"><sup>38</sup></a> return if true the values defined after <code>~</code>. If none apply, the function will return the <code>.default</code> for this element.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Which is only by chance called “none” here as well.&lt;/p&gt;"><sup>39</sup></a></p>
</div>
<p><br>
The table in <code>limma_list[["results"]]</code> contains now all information we need to further plot and report our results.</p>
</div>
<div id="visualizing-the-limma-results" class="section level2" number="8.3">
<h2>
<span class="header-section-number">8.3</span> Visualizing the limma results<a class="anchor" aria-label="anchor" href="#visualizing-the-limma-results"><i class="fas fa-link"></i></a>
</h2>
<p>The one and only way to represent the data from a differential abundance/expression analysis with two groups is the Volcano plot. Let’s try it!</p>
<div class="rmdnote">
<p>Represent the <em>limma</em> results <code>estimate</code> and the <code>-log10(p.value)</code> in a volcano plot and color each point by the type of <code>regulation</code>. Use <code>ggplot2</code> for this.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button46" aria-expanded="false" aria-controls="button46">
Expanation
</button>
<div id="button46" class="collapse">
<p><br>
The <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot2::ggplot()</a></code> function builds the foundation of your plot. You can try this function without any further input.</p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You should see a blank canvas. We can start to add our data with the <code>data</code> argument.</p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> </span></code></pre></div>
<p>Wow, nothing changed. Next, we need to explain <code>ggplot</code> which column of our <code>data</code> should be projected on which axis or plot element. This is achieved with the <code>mapping</code> argument which takes the above described links between data and plot within the <code><a href="https://ggplot2.tidyverse.org/reference/aes.html">ggplot2::aes()</a></code><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;aesthetics&lt;/p&gt;"><sup>40</sup></a> function. We can define columns like <code>x = estimate</code> and even modify them in place like <code>y = -log10(p.value)</code>.</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now that the call gets longer, we can make use of the pipe to add our data.</p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Oh, we have some axes now. But <code>ggplot</code> still only knows which data should be represented where, hence the axes, but not how. We want the most basic type, points.</p>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>,</span>
<span>             y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># geom_point is only one of many geom_... layers </span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Nice, each one of our precious peptides represented as one little ugly point. Let’s unravel this. <code><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">ggplot2::geom_point()</a></code> is called a layer and adds the representation for the defined <code>mapping</code> of our data. It is added to the initial plot created by <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot2::ggplot()</a></code> via a <code>+</code>. No <code>%&gt;%</code>. Just a <code>+</code>. We can further modify this layer by setting a transparency with <code>alpha</code> and changing the <code>shape</code>, so that the changes are better visible.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;You can find a list of all shapes here: &lt;a href="https://ggplot2.tidyverse.org/articles/ggplot2-specs.html#sec:shape-spec" class="uri"&gt;https://ggplot2.tidyverse.org/articles/ggplot2-specs.html#sec:shape-spec&lt;/a&gt;&lt;/p&gt;'><sup>41</sup></a></p>
<div class="sourceCode" id="cb94"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>,</span>
<span>             y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># you can check all options in the function documentation </span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">16</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p>Further, we can change the <code>theme</code> to something like <code><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">ggplot2::theme_classic()</a></code>.</p>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>,</span>
<span>             y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">16</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># there are many different theme_... types </span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Looking good! Now let’s color our points. If we would like to color our points all in one color, we could simple define this in the <code>ggplot2::geom_point(color = "grey)</code> layer. But we want the color to depend on the type of regulation as defined by the thresholds before. For this, we need to add a new variable to the <code><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes()</a></code> mapping and define a manual color scale.</p>
<div class="sourceCode" id="cb96"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>,</span>
<span>             y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span>,</span>
<span>             <span class="co"># Define which column to use</span></span>
<span>             color <span class="op">=</span> <span class="va">regulation</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">16</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># Define the values for each data entry</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>none <span class="op">=</span> <span class="st">"grey"</span>,</span>
<span>                                up <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                                down <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we can remove the space between our plotted points and the outer border like this. This is a taste thing.</p>
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>,</span>
<span>             y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span>,</span>
<span>             color <span class="op">=</span> <span class="va">regulation</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">16</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>none <span class="op">=</span> <span class="st">"grey"</span>,</span>
<span>                                up <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                                down <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># You can also define the xlim = c(-3, 3) and ylim = c(0, 10) here </span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>Finally, we can save our plot first and print it then separately.</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"p"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">limma_list</span><span class="op">[[</span><span class="st">"results"</span><span class="op">]</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">estimate</span>,</span>
<span>             y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span>,</span>
<span>             color <span class="op">=</span> <span class="va">regulation</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">16</span>, alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>none <span class="op">=</span> <span class="st">"grey"</span>,</span>
<span>                                up <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                                down <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_cartesian.html">coord_cartesian</a></span><span class="op">(</span>expand <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span><span class="co"># Sometimes you may need force this by print(limma_list[["p"]]) </span></span>
<span><span class="va">limma_list</span><span class="op">[[</span><span class="st">"p"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
</div>
<p><br>
Great, this basically covered most parameters you can change for plotting in R with <code>ggplot2</code>. It may take some time to get comfortable with all functions and get to know all possibilities, but then there are no limits!</p>
</div>
<div id="summary-1" class="section level2" number="8.4">
<h2>
<span class="header-section-number">8.4</span> Summary<a class="anchor" aria-label="anchor" href="#summary-1"><i class="fas fa-link"></i></a>
</h2>
<p>In this chapter, we applied a package <code>limma</code> to analyse our data and represented the results in one plot. <code>limma</code> is one of the more complicated packages to use IMO, but still very useful and a good exercise.</p>
</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="working-with-data.html"><span class="header-section-number">7</span> Working with data</a></div>
<div class="next"><a href="reporting-results.html"><span class="header-section-number">9</span> Reporting results</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#limma-analysis"><span class="header-section-number">8</span> Limma analysis</a></li>
<li><a class="nav-link" href="#running-limma"><span class="header-section-number">8.1</span> Running limma</a></li>
<li><a class="nav-link" href="#filtering-the-limma-output"><span class="header-section-number">8.2</span> Filtering the limma output</a></li>
<li><a class="nav-link" href="#visualizing-the-limma-results"><span class="header-section-number">8.3</span> Visualizing the limma results</a></li>
<li><a class="nav-link" href="#summary-1"><span class="header-section-number">8.4</span> Summary</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/nicohuttmann/embl-bioinfo-pRoteomics/blob/master/02-group_project.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/nicohuttmann/embl-bioinfo-pRoteomics/edit/master/02-group_project.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Data Analysis in R</strong>" was written by Nico Hüttmann. It was last built on 2025-02-06.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
