<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 7 Working with data | Data Analysis in R</title>
<meta name="author" content="Nico Hüttmann">
<meta name="description" content="Finally, it’s happening. We are beginning to work with real data and you can apply all your new knowledge right away.  7.1 Prepare your project Let’s start with opening a fresh R project, making a...">
<meta name="generator" content="bookdown 0.38 with bs4_book()">
<meta property="og:title" content="Chapter 7 Working with data | Data Analysis in R">
<meta property="og:type" content="book">
<meta property="og:description" content="Finally, it’s happening. We are beginning to work with real data and you can apply all your new knowledge right away.  7.1 Prepare your project Let’s start with opening a fresh R project, making a...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 7 Working with data | Data Analysis in R">
<meta name="twitter:description" content="Finally, it’s happening. We are beginning to work with real data and you can apply all your new knowledge right away.  7.1 Prepare your project Let’s start with opening a fresh R project, making a...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.7.0/transition.js"></script><script src="libs/bs3compat-0.7.0/tabs.js"></script><script src="libs/bs3compat-0.7.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Data Analysis in R</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">Preface</a></li>
<li class="book-part">R Basics</li>
<li><a class="" href="r-and-rstudio.html"><span class="header-section-number">1</span> R and RStudio</a></li>
<li><a class="" href="how-does-r-work.html"><span class="header-section-number">2</span> How does R work?</a></li>
<li><a class="" href="r-packages.html"><span class="header-section-number">3</span> R packages</a></li>
<li><a class="" href="the-tidyverse.html"><span class="header-section-number">4</span> The tidyverse</a></li>
<li><a class="" href="coding-style.html"><span class="header-section-number">5</span> Coding style</a></li>
<li class="book-part">Group Work Project</li>
<li><a class="" href="our-project.html"><span class="header-section-number">6</span> Our project</a></li>
<li><a class="active" href="working-with-data.html"><span class="header-section-number">7</span> Working with data</a></li>
<li><a class="" href="limma-analysis.html"><span class="header-section-number">8</span> Limma analysis</a></li>
<li><a class="" href="reporting-results.html"><span class="header-section-number">9</span> Reporting results</a></li>
<li><a class="" href="more-data-exploration.html"><span class="header-section-number">10</span> More data exploration</a></li>
<li class="book-part">Appendix</li>
<li><a class="" href="version-control-with-github.html"><span class="header-section-number">A</span> Version control with GitHub</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/nicohuttmann/embl-bioinfo-pRoteomics">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="working-with-data" class="section level1" number="7">
<h1>
<span class="header-section-number">7</span> Working with data<a class="anchor" aria-label="anchor" href="#working-with-data"><i class="fas fa-link"></i></a>
</h1>
<p>Finally, it’s happening. We are beginning to work with real data and you can apply all your new knowledge right away.</p>
<div id="prepare-your-project" class="section level2" number="7.1">
<h2>
<span class="header-section-number">7.1</span> Prepare your project<a class="anchor" aria-label="anchor" href="#prepare-your-project"><i class="fas fa-link"></i></a>
</h2>
<p>Let’s start with opening a fresh R project, making a nice folder structure and opening our first <strong>script</strong>. Which kind of script is up to you, both <strong>R</strong> and <strong>R Markdown</strong> work equally well, try and find your preference here.</p>
<div class="rmdnote">
<p>Start your project as described above and save your first script under a name like ‘R_group_project’ in the Scripts folder. Choose a name you like.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button4" aria-expanded="false" aria-controls="button4">
Hint
</button>
<div id="button4" class="collapse">
<p><br>
You can choose what kind of script you want to have by clicking File -&gt; New File -&gt; R Script/R Markdown…/etc.</p>
<p>To save it, press the blue button on top of the script that says Save current document.</p>
</div>
<p><br>
Having our first script ready to code, let’s make it nicer. I would suggest to briefly describe what the script contains and start loading the first packages.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># </span></span>
<span><span class="co"># R Group Project</span></span>
<span><span class="co"># Analysis of LC-MS/MS PELSA data to identify kinases </span></span>
<span><span class="co"># </span></span>
<span></span>
<span><span class="co"># Load libraries</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
<p>Leave an empty lines on top, it will look more professional. You can then also use the <code>#</code> or multiple <code>##</code>s in R scripts as comments<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;text which is not evaluated as code&lt;/p&gt;"><sup>21</sup></a> or as headings in your R Markdown script.</p>
</div>
<div id="import-your-data" class="section level2" number="7.2">
<h2>
<span class="header-section-number">7.2</span> Import your data<a class="anchor" aria-label="anchor" href="#import-your-data"><i class="fas fa-link"></i></a>
</h2>
<p>By now, you should have a link to the datasets in Slack. If not, please complain. There should be two files called report.tsv and report.parquet. Both are output files from the <a href="https://github.com/vdemichev/DiaNN">DIA-NN</a> software, a tool that analyzes mass spectrometry data and outputs a table of all identified ions.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;Each ion, also called precursor, represents a peptide and its charge states. Each peptide can have multiple cahrge states and was originally derived from a protein by proteolytic cleavage. All this data stems from multiple samples and replicates which are combined in one table.&lt;/p&gt;"><sup>22</sup></a></p>
<div class="rmdnote">
<p>Download both files and place them in your R project. Then, try and figure out which functions you need to read both files and store them as objects <code>data_raw</code>.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button5" aria-expanded="false" aria-controls="button5">
Hint
</button>
<div id="button5" class="collapse">
<p><br>
Google :)</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button6" aria-expanded="false" aria-controls="button6">
Another hint
</button>
<div id="button6" class="collapse">
<p><br>
Google ‘import R tsv’, or ‘import R parquet’.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button7" aria-expanded="false" aria-controls="button7">
Code
</button>
<div id="button7" class="collapse">
<p><br><img src="https://raw.githubusercontent.com/tidyverse/vroom/main/img/taylor.gif"></p>
<p>This wasn’t me, I swear. It was <a href="https://vroom.r-lib.org/">them</a>. In all seriousness, there are many ways to import your data from <a href="https://r4ds.had.co.nz/data-import.html#compared-to-base-r">base R</a> solutions, to <a href="https://r4ds.hadley.nz/data-import.html">canonical tidyverse functions</a>, that cover different data types to this one, <code>vroom</code>. It made my life much, much easier as it automatically understands the file format.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simply this</span></span>
<span><span class="va">data_raw</span> <span class="op">&lt;-</span> <span class="fu">vroom</span><span class="fu">::</span><span class="fu"><a href="https://vroom.r-lib.org/reference/vroom.html">vroom</a></span><span class="op">(</span><span class="st">"Data/report.tsv"</span><span class="op">)</span></span></code></pre></div>
<p>All other functions to read files begin with <code>read.</code> or <code>read_</code>. Type them in and press Tab.</p>
<p>Another file format I would like to introduce you to is .parquet, from <a href="https://arrow.apache.org/">Apache Arrow</a> with the accompanying <a href="https://arrow.apache.org/docs/r/">Arrow R Package</a>. This one is not compatible with <code>vroom</code> and needs its own function.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_raw</span> <span class="op">&lt;-</span> <span class="fu">arrow</span><span class="fu">::</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html">read_parquet</a></span><span class="op">(</span><span class="st">"Data/report.parquet"</span><span class="op">)</span></span></code></pre></div>
<p>The same data is a tenth of the size and therefore reads and writes much faster. It is also more convenient as .csv files due to the clear format of the columns. I’m only afraid it doesn’t work with Excel.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button8" aria-expanded="false" aria-controls="button8">
Additional info
</button>
<div id="button8" class="collapse">
<p><br>
Both functions also work with URLs. Try it out:<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;I am not sure how they behave with OwnCloud.&lt;/p&gt;"><sup>23</sup></a></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_raw</span> <span class="op">&lt;-</span> <span class="fu">vroom</span><span class="fu">::</span><span class="fu"><a href="https://vroom.r-lib.org/reference/vroom.html">vroom</a></span><span class="op">(</span><span class="st">"https://www.owncloud.de/bioinfo/R/dataset.tsv"</span><span class="op">)</span></span></code></pre></div>
</div>
<p><br>
We will continue with the data from the .parquet file, so you can delete the other object.</p>
</div>
<div id="understand-your-data" class="section level2" number="7.3">
<h2>
<span class="header-section-number">7.3</span> Understand your data<a class="anchor" aria-label="anchor" href="#understand-your-data"><i class="fas fa-link"></i></a>
</h2>
<p>Okay, we got our data in R. For now, this is already better than Excel. Try and open the .tsv file in Excel. But we need to view it. How do we do this again?</p>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button9" aria-expanded="false" aria-controls="button9">
View solution
</button>
<div id="button9" class="collapse">
<p><br>
Same as for the lists, either click on the object in your Global Environment in the top rigth or type in:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">data_raw</span>, title <span class="op">=</span> <span class="st">"wow, we can add a title"</span><span class="op">)</span></span></code></pre></div>
</div>
<p><br>
Great, we have a first impression of our data by scrolling through its columns. Since understanding data you did not generate yourself is hard, we can discuss this more together. If you want to give it a try, here are some <a href="https://github.com/vdemichev/DiaNN?tab=readme-ov-file#basics-of-dia-data-analysis">DIA-NN basics</a> and more importantly the <a href="https://github.com/vdemichev/DiaNN?tab=readme-ov-file#main-output-reference">Main output reference</a>. This explains all column names, have a look at it.</p>
<p>Once we have an idea what we are looking at, let’s see if R provides us with some more summary functions.</p>
<div class="rmdnote">
<p>Which summary functions did we use in the Intro to R on Monday?</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button10" aria-expanded="false" aria-controls="button10">
Hint
</button>
<div id="button10" class="collapse">
<p><br>
Google gave me this nice resource <a href="https://www.r-bloggers.com/2018/11/explore-your-dataset-in-r/" class="uri">https://www.r-bloggers.com/2018/11/explore-your-dataset-in-r/</a>.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button11" aria-expanded="false" aria-controls="button11">
Code
</button>
<div id="button11" class="collapse">
<p><br>
Here are a couple of examples to get an overview of the structure of an R data object.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Column names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">data_raw</span><span class="op">)</span></span>
<span><span class="co"># or </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">data_raw</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Print the top of the data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">data_raw</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Object structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">data_raw</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">data_raw</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This my take some time with big data but gives you some statistics</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">data_raw</span><span class="op">)</span></span></code></pre></div>
</div>
<p><br>
Not all of these are necessary, I simply <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code> the data before I start to look at it with more specific questions.</p>
<div id="data-format" class="section level3" number="7.3.1">
<h3>
<span class="header-section-number">7.3.1</span> Data format<a class="anchor" aria-label="anchor" href="#data-format"><i class="fas fa-link"></i></a>
</h3>
<p>Having a broad idea of our data, we can start and look for the things we are interested in. Most scientific data follows the structure below.</p>
<div class="inline-figure"><img src="https://r4ds.hadley.nz/images/tidy-1.png" style="width:100.0%"></div>
<p>There are variables, in our case precursors (measured peptide ions). Each variable was measured in an observations, our samples. And for each pair, there are values for the different data types. This is called the wide data format. You can read more about this in the <a href="https://r4ds.hadley.nz/data-tidy.html#sec-tidy-data">Tidy data</a> section. Check if this applies to our data!</p>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button12" aria-expanded="false" aria-controls="button12">
base R solution
</button>
<div id="button12" class="collapse">
<p><br>
The column <code>Runs</code> contains sample names which are the observations of the example above. In addition, we have a column <code>Precursor.Id</code>, which are our variables. This means we are not working with data in the wide format, but in the long format.</p>
<p>The <code>tidyr</code> functions <code><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html">tidyr::pivot_wider()</a></code> and <code><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">tidyr::pivot_longer()</a></code> were mentioned before and will allow as to reshape our data from long to wide and back.</p>
</div>
</div>
<div id="data-types" class="section level3" number="7.3.2">
<h3>
<span class="header-section-number">7.3.2</span> Data types<a class="anchor" aria-label="anchor" href="#data-types"><i class="fas fa-link"></i></a>
</h3>
<p>But before we start to do reshape our data, we need to identify and extract the relevant data columns.</p>
<div class="rmdnote">
<p>Output the data types per each column by combining the two functions <code>lapply</code> and <code>typeof</code>.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button13" aria-expanded="false" aria-controls="button13">
Hint
</button>
<div id="button13" class="collapse">
<p><br><code><a href="https://rdrr.io/r/base/lapply.html">?lapply</a></code> and <code><a href="https://rdrr.io/r/base/typeof.html">?typeof</a></code> will help you understand the individual functions. <code>lapply</code> is a useful function when applying the same operation to all elements of a list or vector. In our case, the <code>data_raw</code> is a tibble/data frame that can be seen as a list of columns. Therefore, the specified function will be applied to each column.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button14" aria-expanded="false" aria-controls="button14">
Code
</button>
<div id="button14" class="collapse">
<p><br></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The function FUN will be applied to each column of X</span></span>
<span><span class="co"># Notice that we did not add `()` brackets after `typeof`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">data_raw</span>, FUN <span class="op">=</span> <span class="va">typeof</span><span class="op">)</span></span></code></pre></div>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button15" aria-expanded="false" aria-controls="button15">
Tidyverse code
</button>
<div id="button15" class="collapse">
<p><br>
The <code><a href="https://purrr.tidyverse.org/reference/map.html">purrr::map</a></code> function family replaces base R’s <code>apply</code> functions and are recommended.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># map works the same as lapply</span></span>
<span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">data_raw</span>, <span class="va">typeof</span><span class="op">)</span></span>
<span><span class="co"># Or </span></span>
<span><span class="va">data_raw</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">typeof</span><span class="op">)</span></span></code></pre></div>
<p><a href="https://purrr.tidyverse.org/reference/map.html">These functions</a> are very useful once you start working with more than one dataset of data frame at once.</p>
</div>
<p><br>
We can use this opportunity to include some discussions about how to <a href="https://adv-r.hadley.nz/control-flow.html">Control flow</a> of your code.</p>
<div class="rmdnote">
<p>Reframe the problem above in a <a href="https://adv-r.hadley.nz/control-flow.html#loops">Loop</a>.</p>
</div>
<p>Now we have an idea which columns in our data contain <code>character</code> data, mainly IDs, and <code>numeric</code> data, which are our quantitative values and some q-values.</p>
<div class="rmdnote">
<p>We can combine identifying data types and extracting these columns with a combination of <code><a href="https://dplyr.tidyverse.org/reference/select.html">dplyr::select()</a></code> and <code><a href="https://tidyselect.r-lib.org/reference/where.html">dplyr::where()</a></code>. 1. Select all <code>character</code> columns. 2. Select all <code>numeric</code> columns. 3. Combine useful ID columns with all <code>numeric</code> columns. This will be helpful for an overview of the data as well.</p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button16" aria-expanded="false" aria-controls="button16">
Hint
</button>
<div id="button16" class="collapse">
<p><br>
You can find explanations for both functions online:
<a href="https://dplyr.tidyverse.org/reference/select.html">select</a> and <a href="https://tidyselect.r-lib.org/reference/where.html">where</a></p>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button17" aria-expanded="false" aria-controls="button17">
Code 1
</button>
<div id="button17" class="collapse">
<p><br>
The <code><a href="https://tidyselect.r-lib.org/reference/where.html">where()</a></code> function accepts any kind of function which can be used to test whether a column should be selected or not.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_raw</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html">where</a></span><span class="op">(</span><span class="va">is.character</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We could also use the output of the previous <code>map</code> function to indicate the columns.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create vector of character column names </span></span>
<span><span class="va">names_character</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_lgl</a></span><span class="op">(</span><span class="va">data_raw</span>, <span class="va">is.character</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="co"># Use vector as input for select </span></span>
<span><span class="va">data_raw</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">names_character</span><span class="op">)</span></span></code></pre></div>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button18" aria-expanded="false" aria-controls="button18">
Code 2
</button>
<div id="button18" class="collapse">
<p><br>
All numeric data can be found with <code><a href="https://rdrr.io/r/base/numeric.html">is.numeric()</a></code>. Again, since we provide a function as an object, and do not want it to do something immediately, we leave out the <code>()</code> brackets.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_raw</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html">where</a></span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<button class="btn btn-primary" type="button" data-toggle="collapse" data-target="#button19" aria-expanded="false" aria-controls="button19">
Code 3
</button>
<div id="button19" class="collapse">
<p><br></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_raw</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="co"># Combine the different ways of choosing columns in a vector c()</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Precursor.Id"</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html">where</a></span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># We could also retain some more information</span></span>
<span><span class="va">data_raw</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Precursor.Id"</span>, <span class="st">"Run"</span>, <span class="st">"Protein.Group"</span>, <span class="st">"Protein.Names"</span>, <span class="st">"Genes"</span>  , <span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html">where</a></span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>But to also provide the most basic use case of <code>select</code>, we can simply use the column names as they are or as “strings”. Pressing Tab with the cursor in the <code>select</code> function can be helpful as well.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_raw</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">Precursor.Id</span>, <span class="va">Modified.Sequence</span>, <span class="st">"Run"</span>, <span class="st">"Precursor.Charge"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<p><br>
Let’s keep the remaining time open to explore other <code>dplyr</code> functions. They are listed in the <a href="https://dplyr.tidyverse.org/reference/index.html">Function reference</a> but the most useful ones are:
* <code><a href="https://dplyr.tidyverse.org/reference/select.html">dplyr::select()</a></code>
* <code><a href="https://dplyr.tidyverse.org/reference/filter.html">dplyr::filter()</a></code>
* <code><a href="https://dplyr.tidyverse.org/reference/arrange.html">dplyr::arrange()</a></code>
* <code><a href="https://dplyr.tidyverse.org/reference/pull.html">dplyr::pull()</a></code>
* <code><a href="https://dplyr.tidyverse.org/reference/bind_rows.html">dplyr::bind_rows()</a></code></p>
<ul>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/mutate.html">dplyr::mutate()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/summarise.html">dplyr::summarise()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/rename.html">dplyr::rename()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">dplyr::full_join()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">dplyr::inner_join()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">dplyr::left_join()</a></code></p></li>
<li><p><code><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">dplyr::right_join()</a></code></p></li>
</ul>
</div>
</div>
<div id="tidy-up-your-data" class="section level2" number="7.4">
<h2>
<span class="header-section-number">7.4</span> Tidy up your data<a class="anchor" aria-label="anchor" href="#tidy-up-your-data"><i class="fas fa-link"></i></a>
</h2>
<p><strong>Coming on Wednesday</strong></p>
<!-- ```{r} -->
<!-- # Check input -->
<!--   if (all(c(Q.Value,  -->
<!--             PG.Q.Value,  -->
<!--             Lib.Q.Value,  -->
<!--             Lib.PG.Q.Value,  -->
<!--             protein.q,  -->
<!--             gg.q) == 0))  -->
<!--     stop("All your q-value cutoffs are 1, please change them according to your study.") -->
<!--   # Filter precursors by proteotypicity -->
<!--   if (proteotypic.only)  -->
<!--     data_raw_filtered <- data_raw %>%  -->
<!--       filter(Proteotypic != 0) -->
<!--   else  -->
<!--     data_raw_filtered <- data_raw -->
<!--   # Filter precursors by Q-values -->
<!--   data_raw_filtered <- data_raw_filtered %>%  -->
<!--     filter(Q.Value <= Q.Value,  -->
<!--            PG.Q.Value <= PG.Q.Value,  -->
<!--            Lib.Q.Value <= Lib.Q.Value,  -->
<!--            Lib.PG.Q.Value <= Lib.PG.Q.Value) -->
<!--   # Extract quantitative data -->
<!--   # Rename runs  -->
<!--   data_raw_filtered <- data_raw_filtered %>%  -->
<!--     mutate(Run = setNames(names(sample_names), unname(sample_names))[Run]) %>%  -->
<!--     arrange(Run) -->
<!--   # Summarise precursors to modified peptides  -->
<!--   data_quant <- data_raw_filtered %>%   -->
<!--     summarise(!!quant_column := sum(!!rlang::sym(quant_column)),  -->
<!--               .by = c("Run", peptide.id))  -->
<!--   # Add group information  -->
<!--   data_quant <- data_quant %>%  -->
<!--     mutate(Group = setNames(sample_groups, names(sample_names))[Run],  -->
<!--            .after = "Run") %>%  -->
<!--     mutate(n = sum(!!rlang::sym(quant_column) > 0), .by = c(peptide.id, "Group")) %>%  -->
<!--     mutate(p = n / max(n), .by = "Group") %>%  -->
<!--     filter(all(p >= min_val_per_group), .by = peptide.id) %>% ##### Check this  -->
<!--     mutate(n_peptide = length(Run), .by = peptide.id) %>%  -->
<!--     filter(n_peptide == max(n_peptide)) %>%  -->
<!--     pivot_wider(id_cols = "Run",  -->
<!--                 names_from = peptide.id,  -->
<!--                 values_from = quant_column) -->
<!-- ``` -->
</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="our-project.html"><span class="header-section-number">6</span> Our project</a></div>
<div class="next"><a href="limma-analysis.html"><span class="header-section-number">8</span> Limma analysis</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#working-with-data"><span class="header-section-number">7</span> Working with data</a></li>
<li><a class="nav-link" href="#prepare-your-project"><span class="header-section-number">7.1</span> Prepare your project</a></li>
<li><a class="nav-link" href="#import-your-data"><span class="header-section-number">7.2</span> Import your data</a></li>
<li>
<a class="nav-link" href="#understand-your-data"><span class="header-section-number">7.3</span> Understand your data</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#data-format"><span class="header-section-number">7.3.1</span> Data format</a></li>
<li><a class="nav-link" href="#data-types"><span class="header-section-number">7.3.2</span> Data types</a></li>
</ul>
</li>
<li><a class="nav-link" href="#tidy-up-your-data"><span class="header-section-number">7.4</span> Tidy up your data</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/nicohuttmann/embl-bioinfo-pRoteomics/blob/master/02-group_project.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/nicohuttmann/embl-bioinfo-pRoteomics/edit/master/02-group_project.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Data Analysis in R</strong>" was written by Nico Hüttmann. It was last built on 2025-02-04.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
